Fix libgap conversion of large Python integers and Sage integers

cxzhong opened the pull request:
This fix #41048 . Now we first convert use GMP, then directly pass the pointer to ``GAP_MakeObjInt``. It fixes the corner case when ``libgap(int)`` is large.


<!-- ^ Please provide a concise and informative title. -->
<!-- ^ Don't put issue numbers in the title, do this in the PR description below. -->
<!-- ^ For example, instead of "Fixes #12345" use "Introduce new method to calculate 1 + 2". -->
<!-- v Describe your changes below in detail. -->
<!-- v Why is this change required? What problem does it solve? -->
<!-- v If this PR resolves an open issue, please link to it here. For example, "Fixes #12345". -->



### :memo: Checklist

<!-- Put an `x` in all the boxes that apply. -->

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [ ] I have updated the documentation and checked the documentation preview.

### :hourglass: Dependencies

<!-- List all open PRs that this PR logically depends on. For example, -->
<!-- - #12345: short description why this is a dependency -->
<!-- - #34567: ... -->

cxzhong committed:
Fix libgap conversion of large Python integers

This fixes the OverflowError when converting Python integers larger than
C long (e.g., int(10**30)) to GAP objects.

The fix adds support for GAP_MakeObjInt() which can construct GAP integers
from GMP limbs, allowing arbitrarily large Python integers to be converted.

Implementation:
- Added GAP_MakeObjInt declaration to gap_includes.pxd
- Modified make_gap_integer() to try fast path first (GAP_NewObjIntFromInt)
- On OverflowError, convert via GMP using mpz_set_pylong and GAP_MakeObjInt
- Properly handle sign by negating size parameter for negative integers
- Added doctests for large positive and negative integers

Fixes: OverflowError: Python int too large to convert to C long

user202729 commented on src/sage/libs/gap/element.pyx:
```diff
@@ -218,16 +218,41 @@ cdef Obj make_gap_integer(sage_int) except NULL:
+            result = GAP_MakeObjInt(<const UInt*>temp[0]._mp_d, size)
+            return result
+        finally:
+            mpz_clear(temp)
```
* does this really work with large **sage** integer?
* avoid blanket try/finally. in which situation would this block raise?
* how do you know a mpz limb has the same size as a gap limb?
* is there a non-internal function that access the limbs?

cxzhong committed:
Fix libgap large integer conversion to avoid GMP internals

- Replace direct access to GMP internal _mp_d field with official APIs
- Use mpz_export() for converting large integers when limb sizes match
- Use mpz_getlimbn() as fallback when sizeof(mp_limb_t) != sizeof(UInt)
- Use PyMem_Malloc/Free for memory management instead of sig_malloc
- Fix limb order: use -1 (least significant first) for GAP compatibility
- Handle zero, negative numbers, and memory errors correctly
- Add comprehensive doctests for small, medium, large integers
- Test both Python int and Sage Integer types
- All 561 doctests pass including new extensive test coverage

user202729 commented on src/sage/libs/gap/element.pyx:
```diff
@@ -218,16 +219,212 @@ cdef Obj make_gap_integer(sage_int) except NULL:
+
+    Test large integers (GMP path with mpz_export)::
+
+        sage: libgap(2**100)
```
the problem is I believe when a sage integer is passed as the argument to `libgap` then this function is not used at all

cxzhong replied:
I do not understand your meaning.
I tested that it is really transferred to a gap element.
```
sage: n = 2**10000
sage: type(n)
<class 'sage.rings.integer.Integer'>
sage: gap_n = libgap(n)
sage: type(gap_n)
<class 'sage.libs.gap.element.GapElement_Integer'>
sage: type(gap_n.sage())
<class 'sage.rings.integer.Integer'>
```

cxzhong replied:
```
sage: int(2**100)
1267650600228229401496703205376
sage: m=int(2**100)
sage: type(m)
<class 'int'>
sage: libgap(m)
1267650600228229401496703205376
sage: type(m)
<class 'int'>
sage: type(libgap(m))
<class 'sage.libs.gap.element.GapElement_Integer'>
sage: type(libgap(m).sage())
<class 'sage.rings.integer.Integer'>
```

user202729 replied:
`libgap(2**10000)` (when the input is a sage integer) already work before this pull request, so that doctest is not testing this function. (I believe it's currently converting the sage integer to decimal string then libgap.eval it, which works but is not as efficient as converting the limbs directly.)

cxzhong replied:
> `libgap(2**10000)` (when the input is a sage integer) already work before this pull request, so that doctest is not testing this function. (I believe it's currently converting the sage integer to decimal string then libgap.eval it, which works but is not as efficient as converting the limbs directly.)

Yes, I modified the doctests. and completely use the new way.

cxzhong committed:
Delete some doctests and update the inputs clarify

Updated docstring to clarify input types and removed outdated test cases.

cxzhong committed:
Refactor integer conversion tests for Sage integers

Updated tests for integer conversion to focus on Sage integers and removed outdated boundary tests.

cxzhong committed:
Refactor tests for libgap function

cxzhong committed:
Change sign variable type from int to Int

tobiasdiez commented on src/sage/libs/gap/element.pyx:
```diff
@@ -214,20 +215,135 @@ cdef Obj make_gap_record(sage_dict) except NULL:
+
+    Test with Python int (not Sage Integer)::
+
+        sage: libgap(int(0))
```
I feel like these tests would profit immensely from being migrated to pytest. For example in these lines, `0` is first auto-converted to a sage integer, then to a python int, then to a gap int, and the printed result is actually a string. So it is in particular not clear if `libgap` would work with a plain python integer (that was not initiated from a sage integer).

cxzhong replied:
I think we can just add a type test after libgap(int(0))

cxzhong replied:
```
sage: type(libgap(int(0)))
<class 'sage.libs.gap.element.GapElement_Integer'>
```

user202729 replied:
> I feel like these tests would profit immensely from being migrated to pytest.

sage has a syntax for this, `0r` (which gets preparsed to `0`). So the argument doesn't hold.

tobiasdiez replied:
> I think we can just add a type test after libgap(int(0))

I don't mean the type of what `libgap` returns, but the type of `0` in these examples - it's a sage `Integer` because of the preparser that is active in these doctests.





> > I feel like these tests would profit immensely from being migrated to pytest.
> 
> sage has a syntax for this, `0r` (which gets preparsed to `0`). So the argument doesn't hold.

I don't see how this helps. `libgap(int(0r))` or `libgap(0r)` in a doctest is still less clear than `libgap(0)` in a normal python file. Also you are still mostly checking the output of `__repr__` in those doctests

cxzhong replied:
> > I think we can just add a type test after libgap(int(0))
> 
> I don't mean the type of what `libgap` returns, but the type of `0` in these examples - it's a sage `Integer` because of the preparser that is active in these doctests.
> 
> > > I feel like these tests would profit immensely from being migrated to pytest.
> > 
> > 
> > sage has a syntax for this, `0r` (which gets preparsed to `0`). So the argument doesn't hold.
> 
> I don't see how this helps. `libgap(int(0r))` or `libgap(0r)` in a doctest is still less clear than `libgap(0)` in a normal python file. Also you are still mostly checking the output of `__repr__` in those doctests

But I think that if you run ``libgap(int(1000000000))`` and like this, it will be also Sage Integer to Python Integer to GAP Integer. We test the setup that Python integer transfer to GAP integer. If it not raises an error and output the same string. I think it is successful. Or we can add a pure python test.
